sftp/location.go

```diff
@@ -0,0 +1,166 @@
+	}, nil
+}
+
+// Containers returns a slice of the Container interface, a cursor, and an error.
+func (l *location) Containers(prefix, cursor string, count int) ([]stow.Container, string, error) {
+	infos, err := l.sftpClient.ReadDir(".")
+	if err != nil {
+		return nil, "", err
+	}
+
+	sort.Slice(infos, func(i, j int) bool { return infos[i].Name() < infos[j].Name() })
+
+	var cont []stow.Container
+	for i, v := range infos {
+		if !v.IsDir() {
+			continue
+		}
+
+		if !strings.HasPrefix(v.Name(), prefix) {
+			continue
+		}
+
+		if v.Name() <= cursor {
+			continue
+		}
+
+		cont = append(cont, &container{
+			location: l,
+			name:     v.Name(),
+		})
+
+		if len(cont) == count {
+			// Check if any of the remaining entries are directories. If they are, then return
+			// v.Name() as the cursor. This prevents us from serving up an empty page of
+			// containers if all of the entries after v.Name() are files.
+			for j := i + 1; j < len(infos); j++ {
+				if infos[j].IsDir() {
+					return cont, v.Name(), nil
+				}
+			}
+			return cont, "", nil
+		}
+	}
+
+	return cont, "", nil
+}
+
+// Close closes the underlying sftp/ssh connections.
+func (l *location) Close() error {
+	var errs error
+
+	if l.sftpClient != nil {
+		if err := l.sftpClient.Close(); err != nil {
+			errs = multierror.Append(errs, errors.Wrap(err, "closing sftp conn"))
+		}
+	}
+
+	if l.sshClient != nil {
+		if err := l.sshClient.Close(); err != nil {
+			errs = multierror.Append(errs, errors.Wrap(err, "closing ssh conn"))
+		}
+	}
+
+	return errs
+}
+
+// Container retrieves a stow.Container based on its name which must be exact.
+func (l *location) Container(id string) (stow.Container, error) {
+	fi, err := l.sftpClient.Stat(id)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, stow.ErrNotFound
+		}
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, stow.ErrNotFound
+	}
+	return &container{
+		location: l,
+		name:     id,
+	}, nil
+}
+
+// RemoveContainer removes a container by name.
+func (l *location) RemoveContainer(id string) error {
+	return recurseRemove(l.sftpClient, id)
+}
+
+// recurseRemove recursively purges content from a path.
+func recurseRemove(client *sftp.Client, path string) error {
+	infos, err := client.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range infos {
+		if !v.IsDir() {
+			return errors.Errorf("directory not empty - %q", v.Name())
+		}
+		if err := recurseRemove(client, filepath.Join(path, v.Name())); err != nil {
+			return err
+		}
+	}
+
+	return client.RemoveDirectory(path)
+}
+
+// ItemByURL retrieves a stow.Item by parsing the URL.
+func (l *location) ItemByURL(u *url.URL) (stow.Item, error) {
+	// expect sftp://<username>@<host>:<port>/<container>/<path to file>
+	// example: sftp://someuser@example.com:22/foo/blah/baz.txt
+
+	urlParts := strings.Split(u.Path, "/")
+	if len(urlParts) < 3 {
+		return nil, errors.New("parsing ItemByURL unexpected length")
+	}
+
+	containerName := urlParts[0]
```

Comment:
Won't the first element of the `u.Path` split always be an empty string?
https://play.golang.org/p/AjanobfQ6Yg

Is this intended?

Reply:
Nice catch. The problem was when I was constructing the item.URL(), we're manually building the url.URL and I wasn't setting the Path on that with a leading slash. Fixing now...